sphinx/builders/__init__.py

```diff
@@ -450,20 +451,29 @@ def read_process(docs: List[str]) -> bytes:
             # allow pickling self to send it back
             return pickle.dumps(self.env, pickle.HIGHEST_PROTOCOL)
 
+        chunks = make_chunks(docnames, nproc)
+
+        chunks_complete = 0
+        summary = bold(__('reading sources... '))
+
         def merge(docs: List[str], otherenv: bytes) -> None:
+            # Update progress message for each chunk completed
+            nonlocal chunks_complete
+            chunks_complete += 1
+            log_status_message(display_chunk(docs), summary, 'purple', chunks_complete,
```

Comment:
How about adding a new option to control the timing of messages (ex. before, after, both) to `status_iterator()` instead? I think the concept of `status_iterator()` is separating progress messages from the main code. But this refactoring makes them coupled. It would be nice if we can keep them less coupled.

Reply:
It's impossible because the for loop doesn't wait until the tasks are completed. The entire loop can finish before even a signle task has.

Reply:
@tk0miya in case I wasn't clear, the `status_iterator` here only measures the progress of _submitting_ tasks into the queue with  `add_task`. `make_chunks` can create more than `nproc` chunks, and `add_task` doesn't block when there are more than `nproc` tasks submitted. It just adds them to a queue ready to be run. So, there is no connection whatsoever between the `status_iterator` progress and actual progress of building the documentation. i.e. the progress messages are pretty much useless.

For writing, there is _some_ serialised work executed inside of the for loop which is measured by the `status_iterator`. However, it's still true that none of the parallelised work is represented in the progress messages. So there can still be huge delays between 100% progress and actually finishing the build step.

Reply:
> separating progress messages from the main code

The logic in the PR looks good to me. Progress messages cannot be decoupled completely from doing the actual work - there has to be some signal that says "chunk is done, please send out a new progress message".

@tk0miya is there a chance this can be merged in the near future? This will be quite helpful for SciPy. We have a very heavy doc build (~20 minutes) that sometimes times out on CI because no output is received.

If there's some extra validation or other activity that we can do to help move this forward, please let us know!